### ISSUE TITLE
Deploy IRI local snapshots

### ISSUE THREAD

**Issue by yillkid** [DLTcollab/general#114]
Deploy IRI local snapshot branch to `node0` and analysis memory consumption.

**Comment by yillkid** [comment]
```
{
    "appName": "IRI",
    "appVersion": "1.6.0-RC12",
    "coordinatorAddress": "KPWCHICGJZXKE9GSUDXZYUAPLHAKAHYHDXNPHENTERYMMBQOPSQIDENXKLKCEYCPVTZQLEEJVYJZV9BWU",
    "duration": 1,
    "features": [
        "snapshotPruning",
        "dnsRefresher",
        "tipSolidification",
        "RemotePOW"
    ],
    "jreAvailableProcessors": 24,
    "jreFreeMemory": 291581632,
    "jreMaxMemory": 1968701440,
    "jreTotalMemory": 811073536,
    "jreVersion": "1.8.0_191",
    "latestMilestone": "KMYOCHNZTSYBNRAESLKRDKYTQMEWUNMYPEBQZW9XSKGS9AVCOYFDVTUZHEEWFVI9ZKOVRQHBIGOK99999",
    "latestMilestoneIndex": 937600,
    "latestSolidSubtangleMilestone": "999999999999999999999999999999999999999999999999999999999999999999999999999999999",
    "latestSolidSubtangleMilestoneIndex": 933210,
    "milestoneStartIndex": 933210,
    "neighbors": 3,
    "packetsQueueSize": 0,
    "time": 1545732616767,
    "tips": 911,
    "transactionsToRequest": 989
}
```

**Comment by jserv** [comment]
Can you summarize what you suffered from?

**Comment by yillkid** [comment]
[Deploy IRI dev-localsnapshots branch (RC12)](https://hackmd.io/s/Hyd4xhkb4)

**Comment by yillkid** [comment]
### Know Issue:

RC12 評估 local snapshot 進入點功能正常，詳見[筆記 - local snapshot without pruning (17:39)](https://hackmd.io/s/Hyd4xhkb4#local-snapshot-without-pruning-1739) log 節錄所示，但  RC12 `LOCAL_SNAPSHOTS_PRUNING_DELAY` 最小值為 50k，使得觸發下一次的 local snapshot 日期過長（如下圖所示），正在評估 patch RC12 或是 downgrade IRI 版本。

![pruning-2](https://user-images.githubusercontent.com/185872/50604955-b0f5f500-0efb-11e9-8a90-1e30682509a9.png)

**Comment by yillkid** [comment]
依照 `LOCAL_SNAPSHOTS_PRUNING_DELAY` 預設最小值 `50000` 的設定，下次執行 local snapshot 的時間點在於 `2019-01-28`，在此之前無法做任何的實驗。

如果要加快測試速度，可以考慮 patch IRI RC12 `LOCAL_SNAPSHOTS_PRUNING_DELAY` 參數的預設最小值，將設定為小於 `50000`，優缺點如下：

### 優點：
* 能夠加快 local snapshot 的測試時間
* 能夠加快測試 local snapshot 執行時候的時間成本

#### 缺點：
* 正常的 `LOCAL_SNAPSHOTS_PRUNING_DELAY=50000` 刪減資料庫後的大小約為 60 GB，此數值與 IF 計算過後的可信賴確認度有關係，任意修改 `LOCAL_SNAPSHOTS_PRUNING_DELAY` 值會導致確認度不可信賴

#### TODO
需要在 Discord 與 IF 開發人員討論

**Comment by yillkid** [comment]
We can modify [LOCAL_SNAPSHOTS_PRUNING_DELAY](https://github.com/iotaledger/iri/blob/dev-localsnapshots/src/main/java/com/iota/iri/conf/BaseIotaConfig.java#L808) and [LOCAL_SNAPSHOTS_PRUNING_DELAY_MIN](https://github.com/iotaledger/iri/blob/dev-localsnapshots/src/main/java/com/iota/iri/conf/BaseIotaConfig.java#L809) default value to speed up local snapshot coming. But we will get a trustless Tangle (Therefore we can't stop original local snapshot IRI).

The other hand, I draw a use scenario to describe how to verify DID claims with the sub-Tangle after puning.

![pruning-3](https://user-images.githubusercontent.com/185872/50739978-802afe00-1222-11e9-8f3c-d924659e31bf.png)

**Comment by yillkid** [comment]
It's different  between local snapshot document and IRI implement, and the newest architecture as below:

#### Architecture:
![](https://i.imgur.com/P9xqQNm.png)

#### Pruning duration test:
(Duration will very dynamic if the SPAM occur!)
| milestone start | milestone end | duration (sec) |
|:------ |:----------- |:----------- |
| 937074 | 937075 | 0.016 |
| 937075 | 937076 | 0.01 |
| 937076 | 937077 | 0.012 |
| 937077 | 937078 | 0.003 |
| 937078 | 937079 | 0.004 |
| 940489 | 940490 | 0.004 |
| 940490 | 940491 | 0.001 |
| 940491 | 940492 | 0.002 |
| 940492 | 940493 | 0.004 |

Refer the [full note on HackMD](https://hackmd.io/s/Hyd4xhkb4) for detail.

#### TODO:
- [ ] Export local snapshot file

**Comment by wusyong** [comment]
I think we can use [Snapshot.ixi](https://github.com/iotaledger/Snapshot.ixi) to export the metadata. This is used when global snapshot taken to make everyone can participate to validate afaik.

**Comment by wusyong** [comment]
InitialSnapshotInex will be [modified](https://github.com/iotaledger/iri/blob/dev/src/main/java/com/iota/iri/service/snapshot/impl/SnapshotProviderImpl.java#L187) whenever a new local snapshot is taken.
So it's slightly different from what it states but not that far.
Official documentation can be found at [here](https://docs.iota.org/iri/usage/local-snapshot).

**Comment by jserv** [comment]
Since IRI `v1.6`, local snapshot is officially integrated. Let's concentrate on the adoption and verification.

**Comment by yillkid** [comment]
1.6.0 running on `node0` already:
Highlighted some important different with RC 12, [note (WIP)](https://hackmd.io/3PGzFzEDQ_ejTW_iZnwQPw).

```
{
    "appName": "IRI",
    "appVersion": "1.6.0-RELEASE",
    "coordinatorAddress": "KPWCHICGJZXKE9GSUDXZYUAPLHAKAHYHDXNPHENTERYMMBQOPSQIDENXKLKCEYCPVTZQLEEJVYJZV9BWU",
    "duration": 21,
    "features": [
        "snapshotPruning",
        "dnsRefresher",
        "tipSolidification",
        "RemotePOW"
    ],
    "jreAvailableProcessors": 24,
    "jreFreeMemory": 151990944,
    "jreMaxMemory": 1968701440,
    "jreTotalMemory": 684720128,
    "jreVersion": "1.8.0_191",
    "lastSnapshottedMilestoneIndex": 933210,
    "latestMilestone": "VYJPGFQCFFXMY9ZOMMWOQATDNK9PRNFYMRPJBVOLDSGQPGYDMQTII9EGMIHGELHZMRXELIMLTZZM99999",
    "latestMilestoneIndex": 965157,
    "latestSolidSubtangleMilestone": "999999999999999999999999999999999999999999999999999999999999999999999999999999999",
    "latestSolidSubtangleMilestoneIndex": 933210,
    "milestoneStartIndex": 960277,
    "neighbors": 3,
    "packetsQueueSize": 0,
    "time": 1547401909100,
    "tips": 619,
    "transactionsToRequest": 1
}
```

**Comment by jserv** [comment]
Close as we work on `v1.6`.
